File: src/Starter/Features/InternalSearchEngines/Settings/ViewModels/KeyboardShortcutInputViewModel.fs
namespace Starter.Features.InternalSearchEngines.Settings.ViewModels

open System
open System.Collections.Generic
open Avalonia.Input
open ReactiveUI
open Starter.Features.Logging
open Starter.Features.Config

type KeyboardShortcutInputViewModel(initialKeyboardShortcut: KeyboardShortcut, onKeyboardChanged: KeyboardShortcut -> unit) as this =
    inherit ReactiveObject() // TODO: Switch to CommunityToolkit

    let mutable listenKeys = false
    let stoppedListening = Event<unit>()
    let stoppedListeningEvent = stoppedListening.Publish

    let mutable keyboardShortcut = initialKeyboardShortcut
    let mutable pressedModifiers = HashSet<Key>()
    let mutable pressedKey = Key.None
    let mutable text = ""

    do this.RefreshText
        keyboardShortcut.Modifiers
        keyboardShortcut.Key

    // Fields
    member this.Text
        with get () = text
        and set v = this.RaiseAndSetIfChanged(&text, v) |> ignore

    member this.RefreshText modifiers key =
        this.Text <-
            modifiers
            |> Seq.map string
            |> fun s -> String.Join(" + ", s)
            |> fun s -> if key <> Key.None then $"{s} + {key}" else s

    member this.StoppedListeningKeys = stoppedListeningEvent

    // Methods
    member this.StartListeningKeys() = listenKeys <- true

    member this.Cancel() =
        listenKeys <- false
        stoppedListening.Trigger()

        pressedKey <- Key.None
        pressedModifiers.Clear()
        this.RefreshText keyboardShortcut.Modifiers keyboardShortcut.Key

    member this.Validate() =
        match pressedModifiers |> Seq.isEmpty, pressedKey with
        | true, _
        | _, Key.None -> this.Cancel()
        | _ ->
            keyboardShortcut <-
                { Modifiers = pressedModifiers |> Seq.toArray
                  Key = pressedKey }

            logger.Information $"Keyboard shortcut changed: %A{keyboardShortcut}"
            this.RefreshText keyboardShortcut.Modifiers keyboardShortcut.Key
            onKeyboardChanged keyboardShortcut

            listenKeys <- false
            stoppedListening.Trigger()

    member this.KeyboardShortcutKeyDown(key: Key) =
        if not listenKeys then () else
        match key with
        | Key.Escape -> this.Cancel()
        | Key.Enter
        | Key.Return -> this.Validate()

        // Modifiers
        | Key.LWin
        | Key.RWin
        | Key.LeftAlt
        | Key.RightAlt
        | Key.LeftCtrl
        | Key.RightCtrl
        | Key.LeftShift
        | Key.RightShift ->
            key |> pressedModifiers.Add |> ignore
            this.RefreshText (Seq.toArray pressedModifiers) pressedKey

        // Other keys
        | _ ->
            pressedKey <- key
            this.RefreshText (Seq.toArray pressedModifiers) pressedKey

File: src/Starter/ViewModels/KeyboardShortcutInputViewModel.cs
using Serilog;
using Starter.Features.Config;
using Starter.Features.PlatformInterop;
using Key = Avalonia.Input.Key;

namespace Starter.ViewModels;

public partial class KeyboardShortcutInputViewModel : ObservableObject
{
    private static readonly IPlatformInterop Platform = PlatformInterop.GetPlatformInterop();

    private bool listenKeys;
    public event Action? StoppedListening;
    public event Action<KeyboardShortcut>? KeyboardShortcutChanged;

    private KeyboardShortcut keyboardShortcut;
    private readonly HashSet<Key> pressedModifiers = [];
    private Key pressedKey = Key.None;
    [ObservableProperty] private string text = "";

    public static bool Enabled => Platform.HotkeyRegistrable;
    public static bool NotEnabled => !Platform.HotkeyRegistrable;

    public KeyboardShortcutInputViewModel(KeyboardShortcut initialKeyboardShortcut)
    {
        keyboardShortcut = initialKeyboardShortcut;
        ResetText();
    }

    private void ResetText()
    {
        var text = string.Join(" + ", keyboardShortcut.Modifiers.Select(key => key.ToString()));
        Text = keyboardShortcut.Key == Key.None ? text : $"{text} + {keyboardShortcut.Key}";
    }

    private void RefreshText()
    {
        var text = string.Join(" + ", pressedModifiers.Select(key => key.ToString()));
        Text = pressedKey == Key.None ? text : $"{text} + {pressedKey}";
    }

    public void StartListening() => listenKeys = true;
    private void StopListening()
    {
        listenKeys = false;
        StoppedListening?.Invoke();
    }

    public void Cancel()
    {
        pressedKey = Key.None;
        pressedModifiers.Clear();
        ResetText();
        StopListening();
    }

    public void Validate()
    {
        if (pressedModifiers.Count == 0 || pressedKey == Key.None)
        {
            Cancel();
            return;
        }

        keyboardShortcut = new KeyboardShortcut(pressedModifiers.ToArray(), pressedKey);
        Log.Information("Keyboard shortcut changed: %A{KeyboardShortcut}", keyboardShortcut);
        RefreshText();
        KeyboardShortcutChanged?.Invoke(keyboardShortcut);

        StopListening();
    }

    public void OnKeyDown(Key key)
    {
        if (!listenKeys) return;
        switch (key)
        {
            case Key.Escape: Cancel(); break;
            case Key.Enter: Validate(); break;

            // Modifiers
            case Key.LWin:
            case Key.RWin:
            case Key.LeftAlt:
            case Key.RightAlt:
            case Key.LeftCtrl:
            case Key.RightCtrl:
            case Key.LeftShift:
            case Key.RightShift:
                pressedModifiers.Add(key);
                RefreshText();
                break;

            // Other keys
            default:
                pressedKey = key;
                RefreshText();
                break;
        }
    }
}
